src/nvim/diff.c

```diff
@@ -358,7 +346,7 @@ static void diff_mark_adjust_tp(tabpage_T *tp, int idx, linenr_T line1, linenr_T
     if (last >= line1 - 1) {
       // 6. change below line2: only adjust for amount_after; also when
       // "deleted" became zero when deleted all lines between two diffs.
-      if (dp->df_lnum[idx] - (deleted + inserted != 0) > line2) {
+      if (dp->df_lnum[idx] - (deleted + inserted != 0) > line2 - dp->is_linematched) {
```

Comment:
yes, I'm effectively converting the greater than symbol to a greater than or equal to.

This is what keeps the diff hunks intact when they are being adjusted in the middle of a diffput / diffget operation that spans multiple diff blocks. It would consider the diff block to be inside a change instead of below it.

So for example you have a diff block on line 3 with a length of 3, and right above this block, 3 new lines get inserted during a 'diffget' operation. that diff block should now be on line 6 with a length of 3, but without that change, it would say the diff block is still on line 3, but now it has a length of 6. That breaks everything and makes the diff  blocks incorrect, so the wrong text would get put / obtained with the diffget / diffput commands. This is what was happening when the diffget / diffput operations were not working initially.

Reply:
and I'm getting the boolean from the diff block because not all diff blocks are linematched. If they are too long, then they would not be linematched, and the diffget / diffput logic doesn't need to change.